lib/gc/Transforms/LowerTileVectorPass.cpp

```diff
@@ -0,0 +1,680 @@
+      reshapeHelper.srcVectorizedShape.emplace_back(cur);
+    }
+
+    if (isCollapseOp and reshapeHelper.shapeScales.count(resultIdx)) {
+      reshapeHelper.srcVectorizedShape.back() *=
+          reshapeHelper.shapeScales[resultIdx];
+    }
+
+    if (!isCollapseOp and reshapeHelper.shapeScales.count(srcIdx)) {
+      reshapeHelper.srcVectorizedShape.back() *=
+          reshapeHelper.shapeScales[srcIdx];
+    }
+
+    cur = 1;
+    resultIdx++;
+  }
+}
+
+/// Need to check whether the reassociation, input, output and input vectorize
+/// size are valid.
+template <class T, typename = typename std::enable_if<
+                       decay_equiv<T, tensor::ExpandShapeOp>::value ||
+                           decay_equiv<T, tensor::CollapseShapeOp>::value,
+                       T>>
+LogicalResult
+lowerReshapeOpPrecondition(T reshapeOp,
+                           ArrayRef<int64_t> inputVectorSizes = {}) {
+
+  Type resultType = reshapeOp->getResultTypes()[0];
+  auto resultShapeType = cast<ShapedType>(resultType);
+  RankedTensorType srcShapeType = reshapeOp.getSrcType();
+
+  // check reassociation
+  SmallVector<int64_t> associateIndices;
+
+  for (const Attribute &attr : reshapeOp.getReassociation()) {
+    llvm::transform(
+        cast<ArrayAttr>(attr), std::back_inserter(associateIndices),
+        [](Attribute indice) { return cast<IntegerAttr>(indice).getInt(); });
+  }
+  if (isContainsDynamicSize(associateIndices)) {
+    LDBG("Reassociation must be static: " << reshapeOp << "\n");
+    return failure();
+  }
+
+  // check input and output shape
+  bool isStaticInputOutput =
+      resultShapeType.hasStaticShape() && srcShapeType.hasStaticShape();
+  if (!isStaticInputOutput) {
+    LDBG("Input and output shape must be static: " << reshapeOp << "\n");
+    return failure();
+  }
+  // ensure specify input vector size is valid
+  if (!inputVectorSizes.empty() &&
+      failed(vector::isValidMaskedInputVector(resultShapeType.getShape(),
+                                              inputVectorSizes)))
+    return failure();
+
+  if (!llvm::all_of(llvm::zip(resultShapeType.getShape(), inputVectorSizes),
+                    [](std::tuple<int64_t, int64_t> sizePair) {
+                      int64_t staticSize = std::get<0>(sizePair);
+                      int64_t inputSize = std::get<1>(sizePair);
+                      return inputSize % staticSize == 0;
+                    })) {
+    LDBG("Input vector sizes must be an integer multiple or equal to "
+         "space static sizes");
+    return failure();
+  }
+
+  return success();
+}
+
+LogicalResult
+lowerConcatOpPrecondition(tensor::ConcatOp concatOp,
+                          ArrayRef<int64_t> inputVectorSizes = {}) {
+  if (!inputVectorSizes.empty()) {
+    LDBG("Concat operation does not support specify inputVectorSizes: "
+         << concatOp << "\n");
+  }
+  // check input operand shape type
+  if (not llvm::all_of(concatOp.getOperandTypes(), [](Type x) {
```

Comment:
Check any_of shape is dynamic maybe faster than all_of must be static?

Reply:
Sure~ Will change it to anyof!